### PR TITLE
Basic Random based implementation of Scout robot with lumberjack avoidance

### PR DESCRIPTION
**Pull request by Lukx19**
This implementation is able to collect bullets from neutral trees and avoid hitting lumberjack.

PROBLEMS:
lumberjacks and other units can make scout go to corners. The scout can be easily killed here because it takes to scout too long to find exit route by random chance.

Possible solution:
Create something like danger map with higher numbers representing more danger.

### CODE REVIEW

File: src/blueteam/Scout.java [hunk 1]
```diff
@@ -1,18 +1,125 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.Random;
+
 
 public class Scout extends Robot {
 
+	Direction dir;
+	Team enemy;
```
Comment (hrnr) [inline src/blueteam/Scout.java]: we have this field in Robot.

File: src/blueteam/Scout.java [hunk 2]
```diff
@@ -1,18 +1,125 @@
 package blueteam;
 
-import battlecode.common.GameActionException;
-import battlecode.common.RobotController;
+import battlecode.common.*;
+
+import java.util.ArrayList;
+import java.util.Arrays;
```
Comment (hrnr) [inline src/blueteam/Scout.java]: unused import

File: src/blueteam/Scout.java [hunk 3]
```diff
@@ -1,18 +1,125 @@
...
+import java.util.Optional;
+import java.util.Random;
+
 
 public class Scout extends Robot {
 
+	Direction dir;
+	Team enemy;
+	Random rand;
```
Comment (hrnr) [inline src/blueteam/Scout.java]: this might want to push this to Robot too.

File: src/blueteam/Scout.java [hunk 4]
```diff
@@ -1,18 +1,125 @@
...
+
+	boolean shakeTheTree(TreeInfo tree) throws GameActionException{
+		if(rc.canShake(tree.getLocation())){
+			rc.shake(tree.getLocation());
+			return true;
+		}
+		return false;
+	}
+
+	Optional<RobotInfo> nearLumberjack(){
```
Comment (hrnr) [inline src/blueteam/Scout.java]: we might want to push these to Robot too, probably make something like getNearestRobot(type, maxRadius) to make it more useful.

File: src/blueteam/Scout.java [hunk 5]
```diff
@@ -1,18 +1,125 @@
...
+			if(!moveInDir(dir))
+				dir = dir.rotateRightDegrees(rand.nextInt(ROTATE_RANGE) + ROTATE_MIN);
+
+		} catch (GameActionException e) {
+			System.out.println("Scout Exception");
+			e.printStackTrace();
+		}
+	}
+
+	boolean moveInDir(Direction dir) throws GameActionException{
```
Comment (hrnr) [inline src/blueteam/Scout.java]: you might want to use a more sophisticated tryMove() from Robot. or we should push this to robot too.

File: src/blueteam/Scout.java [hunk 6]
```diff
@@ -1,18 +1,125 @@
...
+		if(rc.canMove(dir)){
+			rc.move(dir,1);
+			return true;
+		}else{
+			return false;
+		}
+	}
 
+	ArrayList<RobotInfo> filterByType(RobotInfo[] robots, RobotType type){
```
Comment (hrnr) [inline src/blueteam/Scout.java]: this could be one liner: Arrays.stream(robots).filter(x -> x.getType() == type).collect(Collectors.toList()) 
:)

**Review by hrnr — CHANGES_REQUESTED** [review]
please auto-format the code. Also you might want to consider avoiding also other robots (soldiers especially) or just all enemy robots as scouts are fragile.

**Comment by Lukx19** [comment]
The avoidance of other robots will  be added in next round of development. This code avoids only lumberjacks because they kill me in one shot. We should develop dodging which should be used for all our robots.

**Comment by hrnr** [comment]
dodging is already implemented, see #3. I'm anly wandering about lumberjacks, because you are filtering for them explicitly.

File: src/blueteam/Scout.java (hunk 6 above)
Comment (Lukx19) [reply src/blueteam/Scout.java]: The stream doesn't work well in this framework. They take too much byte code. I had it implemented this way and I had to change it :( 
http://archive.battlecodeforum.org/2016/t/problem-using-java-streams/162/6

**Review by Formicarufa — APPROVED** [review]

**Review by hrnr — APPROVED** [review]

**Pull request merged**